### ISSUE TITLE
Proposal: SWIG generator flag to warn about non-declared system types

### ISSUE THREAD

**Issue by luav** [swig/swig#1054]
## Foreword
SWIG generator warns about most of the STL types that has been included into the C++ source files without being declared in the SWIG interface file. This feature allows to see what `std::` types/functions are omitted and **makes the wrapping process usable**.

## Problem
However, **some system types are implicitly omitted without any warnings**, which could cause **non-trivial runtime errors** in the working wrappers (see #1053). For example, there are no any warnings yielded by the SWIG generator for the omitted `uint32_t` from `cstdint` and the SWIG documentation (neither [STL](http://swig.org/Doc3.0/Library.html) nor the [remained sections](http://swig.org/Doc3.0/Contents.html)) also does not mention about the necessity to include `stdint.i`. Moreover, even the wrapper inspection by the non-SWIG developer does not provide any obvious hints about the problem: `SWIGTYPE_p_uint32_t` argument type is generated for the omitted `uint32_t`, `SWIG_AsVal_unsigned_SS_int` for the resolved one by `%include <stdint.i>`. Both types looks like the valid resolved types.

## Proposal
1. It would be nice to have *explicit description* in both [STL section](http://swig.org/Doc3.0/Library.html) and [SWIG Basics](http://swig.org/Doc3.0/SWIG.html) documentation that `SWIG_X` are internally resolved [fundamental  types](http://en.cppreference.com/w/cpp/language/types) and `SWIGTYPE_p_X` are either resolved compound types or unresolved at all...
1.  It would be **very helpful** to have distinct naming of the resolved compound types and non-resolved types, which could be implemented holding additional flag in the types description table and adding additional prefix after the `SWIGTYPE_` or adding some suffix. Now the look identically: `SWIGTYPE_p_OutputOptions` is resolved (`%include` provided) and `SWIGTYPE_p_uint32_t` is not resolved.
1. It would be more usable and helpful to have *additional flag for the SWIG generator* to warn about the system includes (`#include <XXX>`), at least for the headers without the file extension and staring with 'c' (most of the system C includes in C++ include style, for example `#include <cstdint>`). The *flag could be optional and turned on by default* because it might be very helpful, but the users might not want to see lots of warning about all (`#include <XXX.h>`).

**Comment by vadz** [comment]
I think that it would be quite useful to have an option forbidding the generation of `SWIGTYPE_xxx` entirely. It would, of course, need to be off by default to avoid breaking tons of existing projects using SWIG, but I believe it would be useful for the new projects (and some of the existing ones, such as mine, where I typically have a postprocessing step in the makefile, grepping the files produced by SWIG for `SWIGTYPE_` and failing the build if any are found).

**Comment by luav** [comment]
Yes, it also would be a nice feature, n.4 in addition to the n.3!

**Comment by ojwb** [comment]
A way to forbid the opaque `SWIGTYPE` wrappers would certainly be useful, though seems mostly orthogonal to what's proposed here (which seems to be warning about header includes).

Calling these `warn<N>` seems unhelpful - essentially nobody will be able to remember the meanings of five or so different arbitrarily numbered settings for something they don't regularly use, so everyone will have to consult the docs every time.

I'm not really clear how this warning is expected to work either.  SWIG doesn't follow `#include` by default (for sound reasons), which means if I wrap `<somelib.h>` and that includes `somelib/_something.h` which does `#include <cstdint>`, then SWIG doesn't see that `<cstdint>` is in effect.  So in general (and in many real cases) these warnings aren't going to be reliable.

Probably a more workable approach would be to give SWIG a list of type names to warn about if they turn up as opaque types, along with a mapping to what to include to solve that so the warning can point the user to that.  E.g. `uint32_t` -> `%include <stdint.i>`, `DWORD` -> `%include <windows.i>`, etc.  We can probably just generate this mapping automatically (which would be better than having that information in two places, which will inevitably get out of step).

For types in `<stdint.h>`/`<cstdint>` I think we should seriously consider just having them on by default.  The main argument I see for not doing so is that these types are available to user code to define differently if the standard header hasn't been included, but I think at this point if you do `typedef double* uint64_t;` then you deserve some pain - after all, these types were standardised 18 years ago for C and 6 years ago for C++.  I think defaulting to assuming the standard meaning is both more helpful, and entirely reasonable provided an explicit `typedef` or `%apply` for the type overrides the default.

**Comment by luav** [comment]
>  SWIG doesn't follow #include by default ...

The list of all user includes to be warned can be efficiently fetched considering that:
1. The warnings are greedy, i.e. if a warning about the missing `%include XXX` is shown then it does not matter what is included inside the `XXX`.
1. It makes sense to expand only the non-system includes (`#include "X"`, but not `#include <X>`) warning about the missed both system and user includes.

The efficient implementation could be the following. Starting from the SWIG interface file, track the list of the expanding non-system `#include` directives, checking them only:
1. inside the specified `%include` files that do not relate to SWIG/STL (or even only for the `%include "X"`, but not for `%include <X>`) and
1. inside all the specified non-system include (`#include "X"`).

The `#include` files not explicitly included into the interface file can be fetched in the iterative way just using `grep '^\s*#include\s*[<"]' <expanding_incfiles...>` and expanding the non-traversed non-system includes, composing the list of all includes.  
Warnings are yielding for the items in the fetched list of includes, which are not present in the interface file, respecting the specified `warn` strategy.

> Probably a more workable approach would be to give SWIG a list of type names to warn about if they turn up as opaque types, along with a mapping to what to include to solve that. ... We can probably just generate this mapping automatically.

It would be awesome if could be generated automatically! And hardly usable without the automatic generation.  Many of such missed definitions could be present in the mentioned above list of the fetched includes. This list still might be useful as a hint for the user because most of the opaque types usually are not present in the supplied "*.i" files.

For the `<cstdint>` it worth at least to show the warning if it is present in the "#include", but not in the "%include" (if not supported by SWIG out of the box).

**Comment by ojwb** [comment]
The only real difference between `#include <foo.h>` and `#include "foo.h"` is typically that the latter looks in the current directory by default.  So you can't use the quote type to decide "is this a system header", as people can (and do) write things like `#include "stdio.h"` - e.g.:

https://codesearch.debian.net/search?q=%23include+%22stdio.h%22&perpkg=1

And in the other direction, the GNU autoconf manual recommends using `#include <config.h>` to include the configure-generated header containing macros for the results of what configure probed.

And it's perfectly reasonable to wrap a header for an installed library, which will typically be in `/usr/include`, so you can't use the directory a header as much of a guide.

Anyway, the big and fairly easy win here seems to be to essentially `%include <stdint.i>` by default (and ensure that an explicit `typedef` or `%apply` overrides that).  @wsfulton Does that sound reasonable to you for 4.0?

**Comment by vadz** [comment]
FWIW I agree that handling standard types by default would be much more helpful than not handling them and I don't see any problems with this in practice.

**Comment by ojwb** [comment]
Actually, I do see an issue with `stdint.i` by default now I think about it again - `stdint.i` violates the SWIG goal of generating wrappers which are portable as the sizes of the types depend on whether `SWIGWORDSIZE64` or `SWIGWORDSIZE32` is defined. With `stdint.i` by default and how `stdint.i` works currently, it would be easy to inadvertently generate non-portable wrappers without realising it.

The C/C++ wrapper layer could use conditional code and effectively generate both variants[*] and select the correct one at compile time, but there's also target language that depends on the type widths for some target languages.  Maybe that's solvable, but it seems like it would need a lot of new typemaps writing across many of the target languages.

---

_[*] Actually it really should be more than two variants as the current "32 vs 64" distinction oversimplifies reality - see https://github.com/swig/swig/issues/1850#issuecomment-1058770739._

**Comment by wsfulton** [comment]
I can't see stdint.i being included by default. Certainly not until the pile of problems that exist with it are resolved and the numerous missing integral types support is added to SWIG. Quite a bit of fundamental work to be done in this space and best discussed in the specific issues on this topic.

**Comment by sjoubert** [comment]
Hi, pinging this issue as I just hit my use case from #2674: I had a return type from a method that was not wrapped and my users then could do nothing with the returned value. (Admittedly a more extensive test suite of my wrapping could have caught this issue).

Any chance of having progress on this?

If the required modifications are not too big, and with some guidance on the design, I might have a bit of time to look into it if it can help.

**Comment by ojwb** [comment]
There's always a chance, but we have a lot of open issues and a finite number of developers with the vast majority of SWIG development being done in people's spare time.

There are several things discussed above, but it seems the part you're interested in is a way to catch inadvertent creation of an opaque type?  I think we could reasonably warn about that by default.  Sometimes wrapping as an opaque type is useful (e.g. consider C's stdio.h where you get a `FILE*` from `fopen()` and all you can do with it is pass it back to other stdio.h functions) but if that's actually what's wanted for a particular case the user could suppress the warning for that type.

If you want to work on it, you probably "just" need to find the right spot in the code to add this warning.  I'm not entirely sure where that would be though.  @wsfulton probably has a better idea than me.

**Comment by vadz** [comment]
FWIW I've been always (since 15+ years...) combining all calls to SWIG in my makefiles with a check if any opaque type wrappers were created and giving an error in this case, e.g.

```make
some-target:
	@$(RM) -f SWIGTYPE_p_*
	$(SWIG) $(SWIGFLAGS) -o $@ $<
	@badout=`ls SWIGTYPE_p_* 2>/dev/null`; \
	[ -z "$$badout" ] || \
	    (rm $@; \
	     echo "ERROR: SWIG processed some classes as opaque pointers: see"; \
	     echo "$$badout" | sed 's/^/\t/'; \
	     echo "       Correct SWIG input to get rid of them and rerun make."; \
	     exit 1)
```

**Comment by ojwb** [comment]
That approach only works for target languages for which SWIG creates a file for each wrapped type (such as Java).  It also requires the user to know they need to do something like that if they aren't expecting opaque type wrappers.

**Comment by vadz** [comment]
Yes, sure, it's far from ideal, but in practice it works fine if Java (or C#) is _one_ of the languages you target because opaque types will usually be generated for all languages. And you could do something similar with `grep` instead of `ls` for the other languages (e.g. Python).

Anyhow, I'd strongly prefer a `-Wopaque` and, ideally, `-Wopaque=error`, option in SWIG itself myself, but this is the best workaround I can propose for now.

**Comment by sjoubert** [comment]
Yes, I was using the `ls` trick when I still had a Java output, which is no longer the case (Python only). And unfortunately `grep` does not work for Python. No class is generated for opaque types in Python. `SWIGTYPE_p_` hits every types in the C++ wrap files including properly wrapped ones. And so far I've not been able to find a grep pattern that works.

I'll try to have a look at generating a warning when I have a moment.

**Comment by sjoubert** [comment]
So, I've taken some time to look into the swig codebase. Let's just say this is quite something.

In short, I got nowhere to even begin understanding the inner workings and architecture.
I'm afraid that without some tips on how to approach things and guidance on where to look I will not be able to do much :/
